### PR TITLE
[Merged by Bors] - Make the `SystemParam` derive macro more flexible

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Currently, the `SystemParam` derive forces you to declare the lifetime parameters `<'w, 's>`, even if you don't use them.
If you don't follow this structure, the error message is quite nasty.

### Example (before):

```rust
#[derive(SystemParam)]
pub struct EventWriter<'w, 's, E: Event> {
    events: ResMut<'w, Events<E>>,
    // The derive forces us to declare the `'s` lifetime even though we don't use it,
    // so we have to add this `PhantomData` to please rustc.
    #[system_param(ignore)]
    _marker: PhantomData<&'s ()>,
}
```

## Solution

* Allow the user to omit either lifetime.
* Emit a descriptive error if any lifetimes used are invalid.

### Example (after):

```rust
#[derive(SystemParam)]
pub struct EventWriter<'w, E: Event> {
    events: ResMut<'w, Events<E>>,
}
```

---

## Changelog

* The `SystemParam` derive is now more flexible, allowing you to omit unused lifetime parameters.

### CODE REVIEW

File: crates/bevy_ecs/src/system/system_param.rs [hunk 1]
```diff
@@ -1564,7 +1560,7 @@ pub mod lifetimeless {
 /// struct GenericParam<'w,'s, T: SystemParam> {
 ///     field: T,
 ///     #[system_param(ignore)]
-///     // Use the lifetimes, as the `SystemParam` derive requires them
+///     // Use the lifetimes, or rustc will get angry.
```
Comment (mockersf) [inline crates/bevy_ecs/src/system/system_param.rs:1563]: please be more explicit on the error than "rustc will get angry"

**Review by mockersf — APPROVED** [review]

**Review by james7132 — APPROVED** [review]
My only concern is that the `'w` and `'s` lifetimes are a bit arbitrary from a normal user's perspective. Is there a way to better document this?

File: crates/bevy_ecs/macros/src/lib.rs [hunk 2]
```diff
@@ -380,9 +380,14 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
...
-                .into_compile_error()
-                .into();
+            return syn::Error::new_spanned(
+                lt,
+                r#"invalid lifetime name: expected `'w` or `'s`
+ 'w -- refers to data stored in the World.
+ 's -- refers to data stored in the SystemParam's state.'"#,
+            )
+            .into_compile_error()
+            .into();
```
Comment (JoJoJet) [inline crates/bevy_ecs/macros/src/lib.rs:390]: @james7132 do you think this is an improvement?

This usage is called out in the docs, but IMO the best way of teaching this is to just let users do it wrong, then have the error tell them the right way of doing it.

Comment (james7132) [reply crates/bevy_ecs/macros/src/lib.rs:390]: Looks much better to me. Thanks!

**Review by IceSentry — APPROVED** [review]
This is really nice, it will clean up a bunch of my SystemParams

**Review by alice-i-cecile — APPROVED** [review]
bors r+

**Comment by devil-ira** [comment]
This PR is really nice!
It is a breaking change because of the removal of the lifetime on `EventWriter`. Could you add a small migration guide?

**Comment by JoJoJet** [comment]
I've added a migration guide.

**Pull request closed without merging**